jax/_src/interpreters/pxla.py

```diff
@@ -2814,14 +2814,21 @@ def _get_and_check_device_assignment(
     arr_device_assignment = list(i._device_assignment)  # type: ignore
     if not devices:
       if first_sharding_info[0] != arr_device_assignment:
-        raise DeviceAssignmentMismatchError([
-            DeviceAssignmentMismatch(*first_sharding_info),
-            DeviceAssignmentMismatch(arr_device_assignment, s_type, source_info)])
+        # Perhaps the sharding info is the same modulo using lowering_only_client
+        if not all((d1.id == d2.id and
+                    (hasattr(d1.client, "lowering_only_client") or hasattr(d2.client, "lowering_only_client")))
```

Comment:
I don't really like this. Just override this check like I said in jax2tf? In the future this shouldn't be needed so I don't want to add things in the very critical piece of code that depends on something temporary.

Reply:
I did think about this. If I override this function then one would have to maintain two nearly identical copies, because we want the behavior to remain the same except for this detail. It would be hard to test that the behavior is exactly the same. I could also factor out just the check in this function into a separate function to override. But in general these kind of overrides make it very non-obvious what is going on.

I think that we will need to make changes to this function anyway to support cross-platform more cleanly. But perhaps for now this change that unblocks is acceptable?

Reply:
> I think that we will need to make changes to this function anyway to support cross-platform more cleanly.

I don't think there will be any changes required to this function for that. But that's a long way out to think about it since the design is not even final yet.

>  If I override this function then one would have to maintain two nearly identical copies, because we want the behavior

No, my proposal was something else. I said to override this check in a very particular spot when you use LoweringOnlyDevices in jax2tf. At that point you enter into a world where this check will always fail, so it is fine to just return the backend and device_assignment you want without doing any of the other stuff. Think of it like a context manager that only affects 1 line of code. I can prototype it for you and tell you what I exactly meant. Also this function only checks the devices not the sharding info.

> It would be hard to test that the behavior is exactly the same

I don't think it would be hard if you choose to go that route. Just run a unit test that execute both the functions and add it to jax2tf and then we can never break it.

> If I override this function then one would have to maintain two nearly identical copies,

Also, if this is temporary, I would prefer maintaining 2 copies and doing the thing you want for jax2tf. We have a lot of precedent in jax for that anyways. This function hardly changes anyways. But I still think the thing that I mentioned in the chat should work out just fine.

So I don't really think adding a hack like this in this function where d1.client is lowering_only_client comes from a jax2tf internal setting is the right choice (especially since this function is executed for all ops in jax). This is basically leaking across abstraction boundary.

Reply:
We have context managers like this: https://github.com/google/jax/blob/main/jax/_src/test_util.py#L213 that we use in tests so I am pretty confident that when you lower with fake_devices (and you only care about lowering), getting rid of all the checks and just returning a backend should work out pretty well.

Again, I can help prototype it :)